code42/py42
# Fix paging when user specified page_size < settings.items_per_page

### Description of Change ###

The way that paging is supposed to work is essentially:
- Get a page of results
- if the number of items returned in that page is less than the the requested page size, we are at the end and can stop getting pages
- otherwise, get the next page

There are two ways the user can set the page size of items to be retrieved:
- `py42.settings.items_per_page` (defaults to 100)
- `page_size` keyword arg on the `get_all()` function. This takes precedence over `items_per_page`.

Our paging logic currently determines whether or not it should continue by _always_ looking at `items_per_page`:
https://github.com/code42/py42/blob/5b693c5cc206c42db5786a5e26696fb372b3f0c0/src/py42/services/util.py#L9

This is a problem because if a `page_size` is specified that is less than `items_per_page` then the first page will always be treated as if it were the last page.

This is particularly problematic for `sdk.detectionlists.departing_employee.get_all()`and `sdk.detectionslists.high_risk_employee.get_all()` because they are exceptions to the default page size due to restrictions set on that service, and they are hard-coded to have a default page size of 100. Since `items_per_page` defaults to 500, you will only get the first page from these methods unless you set `items_per_page` to 100 or lower.

This PR resolves that issue by using `page_size` for that determination if one is supplied and otherwise using `items_per_page`.


### Testing Procedure ###
Adjust `items_per_page` below such that you should retrieve multiple pages of users in order to get all of them in your environment. `page_size` must be less than `items_per_page`.

```python
import py42.sdk
import py42.settings as settings

sdk = py42.sdk.from_local_account("console.us.code42.com", "user", "pass")
settings.items_per_page = 100

# in v1.8.1, this will always only get 1 page
# with this fix it should return `math.ceil(num_users / page_size)` pages.
for page in sdk.users.get_all(page_size=50):
    print("got page") 
```

### PR Checklist ###
Did you remember to do the below?

- [N/A] Add docstrings for any new public parameters / methods / classes
